--- Day 8: Resonant Collinearity ---
You find yourselves on the roof of a top-secret Easter Bunny installation.

While The Historians do their thing, you take a look at the familiar huge antenna. Much to your surprise, it seems to have been reconfigured to emit a signal that makes people 0.1% more likely to buy Easter Bunny brand Imitation Mediocre Chocolate as a Christmas gift! Unthinkable!

Scanning across the city, you find that there are actually many such antennas. Each antenna is tuned to a specific frequency indicated by a single lowercase letter, uppercase letter, or digit. You create a map (your puzzle input) of these antennas. For example:

............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............
The signal only applies its nefarious effect at specific antinodes based on the resonant frequencies of the antennas. In particular, an antinode occurs at any point that is perfectly in line with two antennas of the same frequency - but only when one of the antennas is twice as far away as the other. This means that for any pair of antennas with the same frequency, there are two antinodes, one on either side of them.

So, for these two antennas with frequency a, they create the two antinodes marked with #:

..........
...#......
..........
....a.....
..........
.....a....
..........
......#...
..........
..........
Adding a third antenna with the same frequency creates several more antinodes. It would ideally add four antinodes, but two are off the right side of the map, so instead it adds only two:

..........
...#......
#.........
....a.....
........a.
.....a....
..#.......
......#...
..........
..........
Antennas with different frequencies don't create antinodes; A and a count as different frequencies. However, antinodes can occur at locations that contain antennas. In this diagram, the lone antenna with frequency capital A creates no antinodes but has a lowercase-a-frequency antinode at its location:

..........
...#......
#.........
....a.....
........a.
.....a....
..#.......
......A...
..........
..........
The first example has antennas with two different frequencies, so the antinodes they create look like this, plus an antinode overlapping the topmost A-frequency antenna:

......#....#
...#....0...
....#0....#.
..#....0....
....0....#..
.#....A.....
...#........
#......#....
........A...
.........A..
..........#.
..........#.
Because the topmost A-frequency antenna overlaps with a 0-frequency antinode, there are 14 total unique locations that contain an antinode within the bounds of the map.

Calculate the impact of the signal. How many unique locations within the bounds of the map contain an antinode?

In [ ]:
"""
Main challenge: to find antinodes, for any pair of same-type antennas a and b, calculate their distance in 2d space given by b = a + d. Then the two points of the antinodes are a - d and a + 2d.

So we:
- for each possible antenna identifier (upper case letters, lower case letters, digits), collect all antenna locations in a dict
- for each pair of antennas of the same type, calculate the locations of the antinodes
- for each antinode check if it is inbounds of the grid
- count unique antinode locations (two antinodes in the same location count as one point)

"""

In [7]:
with open('input8.txt') as file:
    data = [line.rstrip() for line in file]
if len(data) == len(data[0]):
    grid_size = len(data)
grid_size

50

In [25]:
antennas = {}
for i, line in enumerate(data):
    for j, char in enumerate(line):
        if char.isalnum():
            antennas.setdefault(char, []).append((i, j))
antennas

{'y': [(0, 5), (4, 4), (7, 2), (13, 15)],
 'p': [(0, 32), (5, 38), (6, 45), (7, 35)],
 'r': [(0, 49), (5, 48), (14, 44)],
 'I': [(1, 8), (24, 49), (33, 41), (34, 46)],
 '4': [(2, 22), (3, 10), (7, 11), (16, 19)],
 's': [(2, 24), (6, 13), (11, 26), (15, 29)],
 '0': [(6, 10), (8, 15), (9, 14), (13, 19)],
 'N': [(6, 20), (9, 24), (10, 13), (23, 43)],
 '1': [(6, 39), (17, 41), (28, 38), (31, 30)],
 't': [(9, 19), (14, 11), (16, 16), (43, 8)],
 'h': [(9, 29), (20, 32), (24, 43), (37, 10)],
 'j': [(11, 6), (13, 9), (16, 38), (18, 5)],
 'H': [(11, 39), (12, 28), (19, 46), (22, 45)],
 'l': [(11, 43), (24, 32), (25, 20), (28, 21)],
 'O': [(11, 46), (12, 45), (17, 47), (20, 48)],
 'q': [(12, 10), (13, 8), (20, 21), (21, 3)],
 'f': [(13, 2), (30, 29), (36, 22), (37, 27)],
 'e': [(13, 6), (16, 11), (17, 3), (18, 2)],
 'k': [(14, 38), (21, 23), (22, 31), (31, 32)],
 'Q': [(14, 41), (15, 26), (21, 43), (22, 46)],
 '6': [(15, 9), (38, 10), (42, 5), (49, 8)],
 'x': [(15, 33), (16, 35), (21, 46)],
 'W'

In [28]:
def extend_antinode(ant_a, ant_b):
    (a_x, a_y), (b_x, b_y) = ant_a, ant_b
    d_x = b_x - a_x
    d_y = b_y - a_y
    anti_x = b_x + d_x
    anti_y = b_y + d_y
    return(anti_x, anti_y)

def inbounds(coord):
    for c in coord:
        if not 0 <= c <= grid_size - 1:
            return False
    return True


In [31]:
extend_antinode((41, 25), (38, 21))

(35, 17)

In [ ]:
antinode_locations = set()
for loc_by_type in antennas.values():
    for ix, ant_a in enumerate(loc_by_type[:-1]):
        for ant_b in loc_by_type[ix+1:]:
            a1 = extend_antinode(ant_a, ant_b)
            a2 = extend_antinode(ant_b, ant_a)
            if inbounds(a1):
                antinode_locations.add(a1)
            if inbounds(a2):
                antinode_locations.add(a2)
len(antinode_locations)

396

Your puzzle answer was 396.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
Watching over your shoulder as you work, one of The Historians asks if you took the effects of resonant harmonics into your calculations.

Whoops!

After updating your model, it turns out that an antinode occurs at any grid position exactly in line with at least two antennas of the same frequency, regardless of distance. This means that some of the new antinodes will occur at the position of each antenna (unless that antenna is the only one of its frequency).

So, these three T-frequency antennas now create many antinodes:

T....#....
...T......
.T....#...
.........#
..#.......
..........
...#......
..........
....#.....
..........
In fact, the three T-frequency antennas are all exactly in line with two antennas, so they are all also antinodes! This brings the total number of antinodes in the above example to 9.

The original example now has 34 antinodes, including the antinodes that appear on every antenna:

##....#....#
.#.#....0...
..#.#0....#.
..##...0....
....0....#..
.#...#A....#
...#..#.....
#....#.#....
..#.....A...
....#....A..
.#........#.
...#......##
Calculate the impact of the signal using this updated model. How many unique locations within the bounds of the map contain an antinode?

In [ ]:
# One question I have after reading the assignment is if the antinodes have to be multiples of the distances between antennas 1 and 2, or can be a fraction as well *as long as the fraction leaves them exactly in line* (so the distance 4,4 will have antinodes at every 1,1 point, while the distance 4,3 will not since it is not divisible without decimals)
# The framing of the assignment "exactly in line" sounds like the answer is the former (find lowest divisor in natural numbers) but there is no example of that in the assignment and the implementation of the latter is simpler. so im gonna start with the latter

In [88]:
def extend_antinode(ant_a, ant_b):
    (a_x, a_y), (b_x, b_y) = ant_a, ant_b
    d_x = b_x - a_x
    d_y = b_y - a_y
    anti_x = b_x + d_x
    anti_y = b_y + d_y

    if inbounds((anti_x, anti_y)):
        if more_nodes := extend_antinode(ant_b, (anti_x, anti_y)):
            return (more_nodes + [(anti_x, anti_y)])
        else:
            return([(anti_x, anti_y)])
    return []

In [59]:
extend_antinode((20,25), (25,18))

[(35, 4), (35, 4), (30, 11)]

In [89]:
antinode_locations = set()
for loc_by_type in antennas.values():
    for ix, ant_a in enumerate(loc_by_type[:-1]):
        for ant_b in loc_by_type[ix+1:]:
            antinodes = extend_antinode(ant_a, ant_b) + extend_antinode(ant_b, ant_a)
            if antinodes:
                antinode_locations.update(antinodes)
len(antinode_locations)

1075

In [87]:
# The answer is too low. Presumably we have to find fractions in whole numbers. Let's do it minimally invasive to existing code:

from math import gcd

antinode_locations = set()
for loc_by_type in antennas.values():
    for ix, ant_a in enumerate(loc_by_type[:-1]):
        for ant_b in loc_by_type[ix+1:]:
            (a_x, a_y), (b_x, b_y) = ant_a, ant_b
            d_x = b_x - a_x
            d_y = b_y - a_y
            gcd_ = gcd(d_x, d_y)
            d_x, d_y = d_x / gcd_, d_y / gcd_   
            first_an = (a_x + d_x, a_y + d_y)

            antinodes = extend_antinode(ant_a, first_an) + extend_antinode(first_an, ant_a) + [ant_a, ant_b]
            if antinodes:
                antinode_locations.update(antinodes)
len(antinode_locations)

1200

In [90]:
# GCD functionality didnt actually give me more answers. Figuring out a bug did: I had (kinda sloppily) added the original antenna coordinates to the list of antinodes in a recursive function call - only that didn't work because the original coordinates were not transferred in the recursion. Oops! Fair punishment for doing something that I knew was sloppy (I have since taken it out of the function def above)

That's the right answer! You are one gold star closer to finding the Chief Historian.

You have completed Day 8! You can [Share] this victory or [Return to Your Advent Calendar].